### PR TITLE
Adding ArangoDB Provider

### LINKED ISSUE #17778: Add Arango hook
**Description**

https://www.arangodb.com/ is a graph db much like Neo4j.
It has official python connector: https://github.com/ArangoDB-Community/python-arango

**Use case / motivation**

Using ArangoHook to interact with Arango.
Using ArangoOperator to run [simple queries](https://docs.python-arango.com/en/main/simple.html).
Possibly also ArangoAQLHook / ArangoAQLOperator to run [AQL queries](https://docs.python-arango.com/en/main/aql.html) - not sure if it require separate hook/operator.

### PR DESCRIPTION
**Pull request by pateash**
closes: #17778 

---
## Description

Adding ArangoDB provider based on Python SDK https://github.com/ArangoDB-Community/python-arango

Users can create their own custom operators leveraging the **ArangoDBHook**  directly 
or building their operator on **AQLOperator** by providing **result_processor**  method,

```
operator = AQLOperator(
    task_id='aql_operator',
    sql="FOR doc IN students " \
        "RETURN doc",
    dag=dag,
    result_processor=lambda cursor: print([document["name"] for document in cursor])
)
```

Sensor can be implemented by SQL

```
sensor = AQLSensor(
    task_id="aql_sensor",
    sql="FOR doc IN students " \
        "FILTER doc.name == 'judy' " \
        "RETURN doc",
    timeout=60,
    poke_interval=10,
    dag=dag,
)
```

### CODE REVIEW

File: airflow/providers/arangodb/example_dags/example_arangodb.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import logging
+from datetime import datetime
+from typing import Any, Optional
+
+from github import GithubException
+
+from airflow import AirflowException
```
Comment (eladkal) [inline airflow/providers/arangodb/example_dags/example_arangodb.py]: Please check the imports. You have 3 unused imports

File: airflow/providers/arangodb/operators/arangodb.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        For more information on how to use this operator, take a look at the guide:
+        :ref:`howto/operator:AQLOperator`
+
+    :param sql: the AQL query to be executed. Can receive a str representing a
+        sql statement
+    :param result_processor: function to further process the Result from ArangoDB
+    :param arangodb_conn_id: Reference to :ref:`ArangoDB connection id <howto/connection:arangodb>`.
+    """
+
+    template_fields: Sequence[str] = ('sql',)
```
Comment (eladkal) [inline airflow/providers/arangodb/operators/arangodb.py]: Should we add also `template_ext`?

File: airflow/providers/arangodb/sensors/arangodb.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    matches the given query in ArangoDB. Example:
+
+    :param collection: Target DB collection.
+    :param query: The query to find the target document.
+    :param arangodb_conn_id: The :ref:`ArangoDB connection id <howto/connection:arangodb>` to use
+        when connecting to ArangoDB.
+    :param arangodb_db: Target ArangoDB name.
+    """
+
+    template_fields: Sequence[str] = ('sql',)
```
Comment (eladkal) [inline airflow/providers/arangodb/sensors/arangodb.py]: Should we add also `template_ext`?

File: airflow/providers/arangodb/hooks/arangodb.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+from typing import Dict, Optional, Any
+
+from arango import ArangoClient as ArangoDBClient, AQLQueryExecuteError
+from arango.result import Result
+
+from airflow import AirflowException
+from airflow.hooks.base import BaseHook
+
+
+class ArangoDBHook(BaseHook):
```
Comment (eladkal) [inline airflow/providers/arangodb/hooks/arangodb.py:29]: Can we expose also `create_database`, `create_collection`, `create_graph` from the python lib?
These are all very simple 1-2 line functions I believe

**Review by eladkal — COMMENTED** [review]
Over all LGTM

nice job @pateash

**Comment by potiuk** [comment]
Can you fix static checks please.

File: airflow/providers/arangodb/operators/arangodb.py (hunk 2 above)
Comment (pateash) [reply airflow/providers/arangodb/operators/arangodb.py]: yeah, make sense 
added

File: airflow/providers/arangodb/sensors/arangodb.py (hunk 3 above)
Comment (pateash) [reply airflow/providers/arangodb/sensors/arangodb.py]: added

**Comment by pateash** [comment]
image

**Comment by pateash** [comment]
image

File: airflow/providers/arangodb/hooks/arangodb.py (hunk 4 above)
Comment (pateash) [reply airflow/providers/arangodb/hooks/arangodb.py:29]: sure, created.

Comment (eladkal) [reply airflow/providers/arangodb/hooks/arangodb.py:29]: Can you add test coverage for them?

Comment (pateash) [reply airflow/providers/arangodb/hooks/arangodb.py:29]: Actually, we don't have any logic. Its just a wrapper.
adding basic coverage anyways.

**Comment by eladkal** [comment]
@potiuk can you take a look at the test failure?
`AssertionError: List of expected installed packages and image content mismatch. Check /home/runner/work/airflow/airflow/scripts/ci/installed_providers.txt file.`

I don't recall that when adding a new provider we need to edit the CI script

**Comment by potiuk** [comment]
Yeah: seems that for some reason it contains all providers:

```
docker run -it ghcr.io/apache/airflow/main/prod/python3.7:23b7d64b40261dcdcf73187464c6f09b67afcc57  bash
Unable to find image 'ghcr.io/apache/airflow/main/prod/python3.7:23b7d64b40261dcdcf73187464c6f09b67afcc57' locally
23b7d64b40261dcdcf73187464c6f09b67afcc57: Pulling from apache/airflow/main/prod/python3.7
c229119241af: Pull complete 
5a3ae98ea812: Pull complete 
d6bab1fc351b: Pull complete 
f9cea33fb9b5: Pull complete 
23c22d6e5b5d: Pull complete 
b21b38d9bc75: Pull complete 
e52ad88eda59: Pull complete 
5938673019d8: Pull complete 
10aec20ab867: Pull complete 
bfa0b2f2703d: Pull complete 
abea59e2f689: Pull complete 
ffd9264d5a4a: Pull complete 
ea7c97498e3e: Pull complete 
4aed0971f3f7: Pull complete 
8f85ceb1d546: Pull complete 
b6132f0f6227: Pull complete 
83d18601cc4f: Pull complete 
88748a7a2d95: Pull complete 
4f4fb700ef54: Pull complete 
Digest: sha256:bf5da3a686feab47684c036de99a492c3d024fabd4e7a3b69ea9d63ce941b8c8
Status: Downloaded newer image for ghcr.io/apache/airflow/main/prod/python3.7:23b7d64b40261dcdcf73187464c6f09b67afcc57

airflow@54c94bf4e3b9:/opt/airflow$ airflow providers list
package_name                              | description                                                                                     | version
==========================================+=================================================================================================+========
apache-airflow-providers-airbyte          | Airbyte https://airbyte.io/                                                                     | 2.1.4  
apache-airflow-providers-alibaba          | Alibaba Cloud integration (including Alibaba Cloud https://www.alibabacloud.com//)              | 1.1.1  
apache-airflow-providers-amazon           | Amazon integration (including Amazon Web Services (AWS) https://aws.amazon.com/)                | 3.2.0  
apache-airflow-providers-apache-beam      | Apache Beam https://beam.apache.org/                                                            | 3.3.0  
apache-airflow-providers-apache-cassandra | Apache Cassandra http://cassandra.apache.org/                                                   | 2.1.3  
apache-airflow-providers-apache-drill     | Apache Drill https://drill.apache.org/                                                          | 1.0.4  
apache-airflow-providers-apache-druid     | Apache Druid https://druid.apache.org/                                                          | 2.3.3  
apache-airflow-providers-apache-hdfs      | Hadoop Distributed File System (HDFS) https://hadoop.apache.org/docs/r1.2.1/hdfsdesign.html     | 2.2.3  
                                          | and WebHDFS https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-hdfs/WebHDFS.html |        
apache-airflow-providers-apache-hive      | Apache Hive https://hive.apache.org/                                                            | 2.3.2  
apache-airflow-providers-apache-kylin     | Apache Kylin https://kylin.apache.org/                                                          | 2.0.4  
apache-airflow-providers-apache-livy      | Apache Livy https://livy.apache.org/                                                            | 2.2.2  
apache-airflow-providers-apache-pig       | Apache Pig https://pig.apache.org/                                                              | 2.0.4  
apache-airflow-providers-apache-pinot     | Apache Pinot https://pinot.apache.org/                                                          | 2.0.4  
apache-airflow-providers-apache-spark     | Apache Spark https://spark.apache.org/                                                          | 2.1.3  
apache-airflow-providers-apache-sqoop     | Apache Sqoop https://sqoop.apache.org/                                                          | 2.1.3  
apache-airflow-providers-arangodb         | ArangoDB https://www.arangodb.com/                                                              | 1.0.0  
apache-airflow-providers-asana            | Asana https://app.asana.com/                                                                    | 1.1.3  
apache-airflow-providers-celery           | Celery http://www.celeryproject.org/                                                            | 2.1.3  
apache-airflow-providers-cloudant         | IBM Cloudant https://www.ibm.com/cloud/cloudant                                                 | 2.0.4  
apache-airflow-providers-cncf-kubernetes  | Kubernetes https://kubernetes.io/                                                               | 3.1.2  
apache-airflow-providers-databricks       | Databricks https://databricks.com/                                                              | 2.5.0  
apache-airflow-providers-datadog          | Datadog https://www.datadoghq.com/                                                              | 2.0.4  
apache-airflow-providers-dbt-cloud        | dbt Cloud https://www.getdbt.com/product/what-is-dbt/)                                          | 1.0.2  
apache-airflow-providers-dingding         | Dingding https://oapi.dingtalk.com/                                                             | 2.0.4  
apache-airflow-providers-discord          | Discord https://discordapp.com/                                                                 | 2.0.4  
apache-airflow-providers-docker           | Docker https://docs.docker.com/install/                                                         | 2.5.2  
apache-airflow-providers-elasticsearch    | Elasticsearch https://www.elastic.co/elasticsearch                                              | 3.0.2  
apache-airflow-providers-exasol           | Exasol https://docs.exasol.com/home.htm                                                         | 2.1.3  
apache-airflow-providers-facebook         | Facebook Ads http://business.facebook.com/                                                      | 2.2.3  
apache-airflow-providers-ftp              | File Transfer Protocol (FTP) https://tools.ietf.org/html/rfc114                                 | 2.1.2  
apache-airflow-providers-github           | Github https://www.github.com/                                                                  | 1.0.3  
apache-airflow-providers-google           | Google services including:                                                                      | 6.7.0  
                                          |                                                                                                 |        
                                          |   - Google Ads https://ads.google.com/                                                          |        
                                          |   - Google Cloud (GCP) https://cloud.google.com/                                                |        
                                          |   - Google Firebase https://firebase.google.com/                                                |        
                                          |   - Google LevelDB https://github.com/google/leveldb/                                           |        
                                          |   - Google Marketing Platform https://marketingplatform.google.com/                             |        
                                          |   - Google Workspace https://workspace.google.pl/ (formerly Google Suite)                       |        
apache-airflow-providers-grpc             | gRPC https://grpc.io/                                                                           | 2.0.4  
apache-airflow-providers-hashicorp        | Hashicorp including Hashicorp Vault https://www.vaultproject.io/                                | 2.1.4  
apache-airflow-providers-http             | Hypertext Transfer Protocol (HTTP) https://www.w3.org/Protocols/                                | 2.1.2  
apache-airflow-providers-imap             | Internet Message Access Protocol (IMAP) https://tools.ietf.org/html/rfc3501                     | 2.2.3  
apache-airflow-providers-influxdb         | InfluxDB https://www.influxdata.com/                                                            | 1.1.3  
apache-airflow-providers-jdbc             | Java Database Connectivity (JDBC) https://docs.oracle.com/javase/8/docs/technotes/guides/jdbc/  | 2.1.3  
apache-airflow-providers-jenkins          | Jenkins https://jenkins.io/                                                                     | 2.0.7  
apache-airflow-providers-jira             | Atlassian Jira https://www.atlassian.com/                                                       | 2.0.4  
apache-airflow-providers-microsoft-azure  | Microsoft Azure https://azure.microsoft.com/                                                    | 3.7.2  
apache-airflow-providers-microsoft-mssql  | Microsoft SQL Server (MSSQL) https://www.microsoft.com/en-us/sql-server/sql-server-downloads    | 2.1.3  
apache-airflow-providers-microsoft-psrp   | This package provides remote execution capabilities via the                                     | 1.1.3  
                                          | PowerShell Remoting Protocol (PSRP)                                                             |        
                                          | https://docs.microsoft.com/en-us/openspecs/windowsprotocols/ms-psrp/                            |        
apache-airflow-providers-microsoft-winrm  | Windows Remote Management (WinRM) https://docs.microsoft.com/en-us/windows/win32/winrm/portal   | 2.0.5  
apache-airflow-providers-mongo            | MongoDB https://www.mongodb.com/what-is-mongodb                                                 | 2.3.3  
apache-airflow-providers-mysql            | MySQL https://www.mysql.com/products/                                                           | 2.2.3  
apache-airflow-providers-neo4j            | Neo4j https://neo4j.com/                                                                        | 2.1.3  
apache-airflow-providers-odbc             | ODBC https://github.com/mkleehammer/pyodbc/wiki                                                 | 2.0.4  
apache-airflow-providers-openfaas         | OpenFaaS https://www.openfaas.com/                                                              | 2.0.3  
apache-airflow-providers-opsgenie         | Opsgenie https://www.opsgenie.com/                                                              | 3.0.3  
apache-airflow-providers-oracle           | Oracle https://www.oracle.com/en/database/                                                      | 2.2.3  
apache-airflow-providers-pagerduty        | Pagerduty https://www.pagerduty.com/                                                            | 2.1.3  
apache-airflow-providers-papermill        | Papermill https://github.com/nteract/papermill                                                  | 2.2.3  
apache-airflow-providers-plexus           | Plexus https://plexus.corescientific.com/                                                       | 2.0.4  
apache-airflow-providers-postgres         | PostgreSQL https://www.postgresql.org/                                                          | 4.1.0  
apache-airflow-providers-presto           | Presto https://prestodb.github.io/                                                              | 2.1.2  
apache-airflow-providers-qubole           | Qubole https://www.qubole.com/                                                                  | 2.1.3  
apache-airflow-providers-redis            | Redis https://redis.io/                                                                         | 2.0.4  
apache-airflow-providers-salesforce       | Salesforce https://www.salesforce.com/                                                          | 3.4.3  
apache-airflow-providers-samba            | Samba https://www.samba.org/                                                                    | 3.0.4  
apache-airflow-providers-segment          | Segment https://segment.com/                                                                    | 2.0.4  
apache-airflow-providers-sendgrid         | Sendgrid https://sendgrid.com/                                                                  | 2.0.4  
apache-airflow-providers-sftp             | SSH File Transfer Protocol (SFTP) https://tools.ietf.org/wg/secsh/draft-ietf-secsh-filexfer/    | 2.5.2  
apache-airflow-providers-singularity      | Singularity https://sylabs.io/guides/latest/user-guide/                                         | 2.0.4  
apache-airflow-providers-slack            | Slack https://slack.com/                                                                        | 4.2.3  
apache-airflow-providers-snowflake        | Snowflake https://www.snowflake.com/                                                            | 2.6.0  
apache-airflow-providers-sqlite           | SQLite https://www.sqlite.org/                                                                  | 2.1.3  
apache-airflow-providers-ssh              | Secure Shell (SSH) https://tools.ietf.org/html/rfc4251                                          | 2.4.3  
apache-airflow-providers-tableau          | Tableau https://www.tableau.com/                                                                | 2.1.7  
apache-airflow-providers-telegram         | Telegram https://telegram.org/                                                                  | 2.0.4  
apache-airflow-providers-trino            | Trino https://trino.io/                                                                         | 2.1.2  
apache-airflow-providers-vertica          | Vertica https://www.vertica.com/                                                                | 2.1.3  
apache-airflow-providers-yandex           | Yandex including Yandex.Cloud https://cloud.yandex.com/                                         | 2.2.3  
apache-airflow-providers-zendesk          | Zendesk https://www.zendesk.com/                                                                | 3.0.3  
```

**Comment by potiuk** [comment]
Let me rebase and see it happening again :)

**Comment by eladkal** [comment]
I don't recall we had such issue when GitHub provider was added (and it was after https://github.com/apache/airflow/commit/621d17bb77e3160c1a927803e5d190c0e2aade3c )

**Comment by potiuk** [comment]
> I don't recall we had such issue when GitHub provider was added (and it was after [621d17b](https://github.com/apache/airflow/commit/621d17bb77e3160c1a927803e5d190c0e2aade3c) )

Me neither. It basicallly SHOUD NOT happen :D. Yet it seems it did again

**Comment by potiuk** [comment]
OK. I know what causes it but I do not know why it happens yet. When PROD build image is prepared we prepare "airflow" package so that it can be installed there from latest sources. But for SOME reason, it contains "all" providers as well. not only airflow. I do not know where it came from yet. But It proves the tests from @mik-laj are useful to catch it.

**Comment by potiuk** [comment]
Still puzzled :) but I am getting closer to solve it

**Comment by pateash** [comment]
> 

thanks @potiuk.

**Comment by pateash** [comment]
> Rebased it @pateash -> I have high hopes for #22649 to either fix it or make it easier to understand where it came from

Sure, let me rebase 🤞

**Comment by potiuk** [comment]
Funny thing - this PR is the only one with the problem. And I already know WHAT happens (when preparing airflow package it includes all the providers where it shoud not), but I have no idea WHY it happens.

**Comment by potiuk** [comment]
@mik-laj KUDOS, KUDOS KUDOS for the test you did .... This is a FANTASTIC  test that saved us from HUGE PROBLEM. If not your test, we would not have discovoered the problem until it would be too late!

**Comment by potiuk** [comment]
> I am not sure, Is it possible to do something with MANIFEST.in in CI?

I do not want to even try to do it or workaround it. There might be many more problems due to that and I do not want to make Airflow workaround such deeply broken problems. @uranusjr might confirm that this is entirely wrong. Arango DB has terribly broken dependencies.

**Comment by joowani** [comment]
Hi maintainer of python-arango here. I've removed the dependency. Please try again with release version [7.3.2](https://github.com/ArangoDB-Community/python-arango/releases/tag/7.3.2). Thanks.

**Comment by potiuk** [comment]
> Hi maintainer of python-arango here. I've removed the dependency. Please try again with release version [7.3.2](https://github.com/ArangoDB-Community/python-arango/releases/tag/7.3.2). Thanks.

Cool. Thanks! @pateash  -> can you add >=7.3.2  to our requirements please ?

**Comment by pateash** [comment]
voila 🥳,
It worked.

**Review by potiuk — APPROVED** [review]
LGTM :) - @eladkal ?

**Comment by eladkal** [comment]
I'll take a look later today

File: docs/apache-airflow/extra-packages-ref.rst [hunk 5]
```diff
@@ -256,6 +256,8 @@ Those are extras that add dependencies needed for integration with other softwar
 +---------------------+-----------------------------------------------------+-------------------------------------------+
 | trino               | ``pip install 'apache-airflow[trino]'``             | All Trino related operators & hooks       |
 +---------------------+-----------------------------------------------------+-------------------------------------------+
+| arangodb            | ``pip install 'apache-airflow[arangodb]'``          | ArangoDB operators, sensors and hook      |
```
Comment (eladkal) [inline docs/apache-airflow/extra-packages-ref.rst:259]: I think this list is sorted alphabetically?

**Review by eladkal — APPROVED** [review]
LGTM

**Comment by potiuk** [comment]
:tada: :tada: :tada: :tada: :tada: :tada: :tada:

**Pull request merged**